wcoss2 cycled ecflow for developer

lgannoaa opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**
In this PR it will contain:
1. The wcoss2 cycled ecflow definition files
2. ecf/include files for developer
3. cycled workflow management jobs

There are 10 files changed in this PR compare to dev_v16:
- ecf/defs/cycled_gfs.def - A new cycled all in one defination files.
- prod00.def, prod06.def, prod12.def, prod18.def - deleted as they are no longer been used in wcoss2 ecflow workflow
- ecf/include/envir-p1.h - The envir-p1.h for developer use that is not wired to NCO production environment
- ecf/scripts/workflow_manager/cycled/workflow_monitor.ecf - Part of the workflow_manager family that is used to control when to start run gfs/gdas/and enkfgdas jobs for the cycle. It check and wait for the required obsproc input available to set the release flag.
- $HOMEgfs/jobs/JECF_MONITOR - JJOB for workflow_monitor to call.
- ecf/scripts/workflow_manager/cycled/next_run.ecf - Part of the workflow_manager family that is used to control family requeue and PDY advance

closes #721 
<!-- Please include relevant motivation and context. -->
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- List any dependencies that are required for this change. -->

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->

**Type of change**

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->

<!-- Use the following as a guide to list your tests and delete options that are not relevant. Expand as necessary. -->
<!--

-->
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

lgannoaa committed:
Replace ecf directory with cycled gfs ecflow workflow for wcoss2

684

lgannoaa committed:
ecflow PR will be submit in 3 parts:
1.
  New cycled definition file
  ecflow scripts from current dev_v16
  New cycled workflow management file
  New archive JJOBs
2.
  Developer enabled ecflow files - (modify the current dev_v16 with developer testing COM assignment)
3.
  Verification jobs.

This is the ecflow PR#1:
  New cycled definition file
  ecflow scripts from current dev_v16
  New cycled workflow management file
  New archive JJOBs

aerorahul commented on ecf/include/emc_cdate.h:
```diff
@@ -0,0 +1,11 @@
+datedir=${DATAROOT}/emc_ecflow_pdy.date.$$.$RANDOM
```
What is the use of this file? And why is this only now needed for archive jobs?

lgannoaa replied:
It is removed. Please do a page reload to see the update.

aerorahul commented on ecf/scripts/workflow_manager/cycled/gdas/jgdas_emc_arch.ecf:
```diff
@@ -0,0 +1,67 @@
+export cyc=%CYC%
+export cycle=t%CYC%z
+export EXPDIR=${EXPDIR:-$HOMEgfs/parm/config}
+%include <emc_cdate.h>
```
Why are these needed?
They don' appear in any of the other j-job cards.

lgannoaa replied:
The vrfy and arch jobs has been removed from this PR because Arun requested it will be in a different PR. And the scope of the vrfy and arch JJobs is now been changed to include modification of rocoto arch.sh and vrfy.sh files pending rocoto workflow test.

aerorahul commented on ecf/scripts/workflow_manager/cycled/gdas/jgdas_emc_arch.ecf:
```diff
@@ -0,0 +1,67 @@
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.nsstbufr
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr_pre-qc
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr.acft_profiles
```
Why are these files being linked in the `COMIN` directory in the archive step?

lgannoaa replied:
The vrfy and arch jobs has been removed from this PR because Arun requested it will be in a different PR. And the scope of the vrfy and arch JJobs is now been changed to include modification of rocoto arch.sh and vrfy.sh files pending rocoto workflow test.

aerorahul commented on ecf/scripts/workflow_manager/cycled/gdas/jgdas_emc_arch.ecf:
```diff
@@ -0,0 +1,67 @@
+cd $ecf_log_dir
+[[ -s gdas_${CDATE}_archive_list.sh ]] && rm -f gdas_${CDATE}_archive_list.sh
+grep "export CDATE=$CDATE" gdas_*|grep -v "Binary file" | awk -F  ":" -v target_n=$archive_log_dir '{print "cp -p",$1,target_n"/"$1".log"}' &> gdas_${CDATE}_archive_list.sh
+sh gdas_${CDATE}_archive_list.sh
```
Why is all this in an `ecf`?  
Is this doing more than the original `arch.sh`? 
If so, please explain.  Or remove.

aerorahul commented on ecf/scripts/workflow_manager/cycled/gfs/jgfs_emc_arch.ecf:
```diff
@@ -0,0 +1,67 @@
+cd $ecf_log_dir
+[[ -s gfs_${CDATE}_archive_list.sh ]] && rm -f gfs_${CDATE}_archive_list.sh
+grep "export CDATE=$CDATE" gfs_*|grep -v "Binary file" | awk -F  ":" -v target_n=$archive_log_dir '{print "cp -p",$1,target_n"/"$1".log"}' &> gfs_${CDATE}_archive_list.sh
+sh gfs_${CDATE}_archive_list.sh
```
same question as for the gdas arch jobcard.

lgannoaa replied:
The vrfy and arch jobs has been removed from this PR because Arun requested it will be in a different PR. And the scope of the vrfy and arch JJobs is now been changed to include modification of rocoto arch.sh and vrfy.sh files pending rocoto workflow test.

aerorahul commented on ecf/scripts/workflow_manager/cycled/gfs/jgfs_emc_arch.ecf:
```diff
@@ -0,0 +1,67 @@
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.nsstbufr
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr_pre-qc
+ln -s ${COMIN_OBS}/${RUN}.${cycle}.prepbufr.acft_profiles
```
same comment as the gdas job card.

aerorahul commented on ecf/scripts/workflow_manager/cycled/workflow_monitor.ecf:
```diff
@@ -0,0 +1,120 @@
+#PBS -S /bin/bash
```
Please explain what this job card is for and why is this now needed and was not before

lgannoaa replied:
I updated the description to include information for each changed file. Include this one. This file is part of the workflow_manager family that is used to control when to start run gfs/gdas/and enkfgdas jobs for the cycle. It check and wait for the required obsproc input available to set the release flag.

aerorahul commented on ecf/scripts/workflow_manager/cycled/next_run.ecf:
```diff
@@ -0,0 +1,59 @@
+export RUN=%RUN%
+export CDUMP=%RUN%
+
+############################################################
+# Load modules
+############################################################
+
+module list
```
What modules are being loaded and listed?

lgannoaa replied:
No need to load module for this job.

aerorahul commented on ecf/scripts/workflow_manager/cycled/next_run.ecf:
```diff
@@ -0,0 +1,59 @@
+
+module list
+
+export edate=%EDATE%
```
Where is %EDATE% defined?  Presumably you would in the `cycled_gfs.def`, but I didn't see it in that file

lgannoaa replied:
The EDATE is added at the beginning of ecflow workflow setup were user modify config.base to set EDATE. The user will enter associated commend for ecflow.

aerorahul replied:
`config.base` defines `EDATE`.  Yes this is true.
I am trying to understand where  `config.base` is being sourced before this point in this particular job to obtain `%EDATE%`.

lgannoaa replied:
At this time. It is not been done automatically. It is done my enter ecflow commend at the time first setup the workflow.

aerorahul replied:
Ok.  In a follow up PR, please add a document `docs/ecflow_dev.md` to outline how a developer could setup and execute an ecflow based workflow.

lgannoaa replied:
ok

aerorahul commented on ecf/scripts/workflow_manager/cycled/next_run.ecf:
```diff
@@ -0,0 +1,59 @@
+  ecflow_client --msg="***Cycled parallel run is completed***"
+else
+  ecflow_client --alter change variable PDY $PDYp1 /cycled_gfs/primary/${cyc}
+  ecflow_client --msg="***Cycled parallel $PDY$cyc is done $PDYp1$cyc is in quote next***"
```
```suggestion
  ecflow_client --msg="***Cycled parallel $PDY$cyc is done $PDYp1$cyc is in queue next***"
```

lgannoaa replied:
Modified

aerorahul commented on ecf/scripts/workflow_manager/cycled/workflow_monitor.ecf:
```diff
@@ -0,0 +1,120 @@
+  if [ $ic -gt 20 ]; then
+    exit 99
+  fi
+  sleep 38
```
There seems some arbitrary sleeping going on here.

lgannoaa replied:
Yes. It is required because the cycled ecflow can be done as fast as 2 hours for each cycle. However, the required NCO obsproc job output may not be available. Hence, implement this delay in workflow_monitor job.

aerorahul commented on ecf/scripts/workflow_manager/cycled/workflow_monitor.ecf:
```diff
@@ -0,0 +1,120 @@
+  sleep 38
+done
+
+#### Ensure MetPlus input exist
+release_fg_metpg2o1="NO"
+release_fg_metppcp1="NO"
+COMIN_vppg="/lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/global"
+COMIN_vppg_gdas=${COMIN_vppg}/archive/obs_data/prepbufr/gdas/prepbufr.gdas.${PDYm1}00
+COMIN_vppg_nam_PDY=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDY}        # /nam.t*z.prepbufr.tm*
+COMIN_vppg_nam_PDYm1=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDYm1}    # /nam.t*z.prepbufr.tm*
+COMIN_vppg_nam_PDYm2=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDYm2}    # /nam.t*z.prepbufr.tm*
+COMIN_vppg_ccpa=${COMIN_vppg}/archive/obs_data/ccpa_accum24hr/ccpa.${PDYm1}12.24h
+echo "Checking metpg2o1 input for prepbufr file on ${PDYm1}"
+ic=1
```
I don't see any Metplus jobs running in the GFS suite. How is this section and below relevant.

lgannoaa replied:
To run METPlus jobs we need to compare files from verification group under /lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/global. 
Without these file, job will silent fail. Hence developed this code to check on input and release job as needed.

lgannoaa replied:
This job has been modified with a new jjob.

aerorahul reviewed (changes requested):
The description of this PR in the description section is inadequate to understand the various changes this PR intends to do.
For e.g. the PR description says what files are added, but provides no information what they do or what they are supposed to be for.
It also does not describe the purpose of the new files that have been added. And there is no information within the file either.
New "job cards" are added, but there is no description of what they are for or why they are needed.

The PR is still too big for review and should be split into different PRs.

I also see new J-Jobs are added for Archive.  These need to be tested for `develop` with rocoto and verified.

See inline comments for specific details.

lgannoaa committed:
Remove archive and verification jobs for a different PR

lgannoaa committed:
remove emc_cdate.h

aerorahul commented on ecf/scripts/workflow_manager/cycled/next_run.ecf:
```diff
@@ -0,0 +1,58 @@
+#PBS -q %QUEUE%
+#PBS -A %PROJ%-%PROJENVIR%
+#PBS -l walltime=00:30:00
+#PBS -l place=vscatter,select=1:ncpus=1:mem=1GB
```
Please put `place` and `select` on separate lines as is the case in all the `ecf` job cards.

lgannoaa replied:
Ok

aerorahul commented on ecf/scripts/workflow_manager/cycled/workflow_monitor.ecf:
```diff
@@ -0,0 +1,119 @@
+export cyc=%CYC%
+export cycle=t${cyc}z
+
+GDATE=$($NDATE -06 $CDATE)
```
Every thing from below this point does not belong in a job card.
A job card is not to be treated like a script.
Its purpose is to call a J-Job.  It seems this needs a J-Job similar to post manager.

lgannoaa replied:
ecf script file is not just a job card. It is a script that contains a job card. There is no such a requirement that ecf script can not contain some scripts. In addition, NCO will not run this script in question.

lgannoaa replied:
@arunchawla-NOAA and @aerorahul, do I need to create a new JJob or can I retain my current solution? I need your guidance.

aerorahul replied:
I would like this in a job card please to conform with existing norms and practices.

lgannoaa replied:
Ok. Will create a new JJob.

lgannoaa replied:
A new JJOB has been created.

lgannoaa commented:
> The description of this PR in the description section is inadequate to understand the various changes this PR intends to do. For e.g. the PR description says what files are added, but provides no information what they do or what they are supposed to be for. It also does not describe the purpose of the new files that have been added. And there is no information within the file either. New "job cards" are added, but there is no description of what they are for or why they are needed.
> 
> The PR is still too big for review and should be split into different PRs.
> 
> I also see new J-Jobs are added for Archive. These need to be tested for `develop` with rocoto and verified.
> 
> See inline comments for specific details.

Hi @aerorahul , please remove this comment because the JJobs, arch, and vrfy job changes have been moved to a different PR.

aerorahul commented:
> > The description of this PR in the description section is inadequate to understand the various changes this PR intends to do. For e.g. the PR description says what files are added, but provides no information what they do or what they are supposed to be for. It also does not describe the purpose of the new files that have been added. And there is no information within the file either. New "job cards" are added, but there is no description of what they are for or why they are needed.
> > The PR is still too big for review and should be split into different PRs.
> > I also see new J-Jobs are added for Archive. These need to be tested for `develop` with rocoto and verified.
> > See inline comments for specific details.
> 
> Hi @aerorahul , please remove this comment because the JJobs, arch, and vrfy job changes have been moved to a different PR.

The comment was made at the time when the j-jobs were included as part of the original PR.  It will remain as part of the review to reflect that they were removed following reviewer comments. Thanks.

aerorahul commented on ecf/scripts/workflow_manager/cycled/next_run.ecf:
```diff
@@ -0,0 +1,58 @@
+export edate=%EDATE%
+export cyc=%CYC%
+export cycle=t${cyc}z
+export EXPDIR=${EXPDIR:-$HOMEgfs/parm/config}
```
So if `EXPDIR` is not defined at this point, it will fill logs in the `HOMEgfs/parm/config`?
Also, how is ecflow receiving this value for `EXPDIR`? I don' see it defined in `cycled_gfs.def`.
Or is it being set in the environment prior to running `ecflow_client` loading of the suite?

lgannoaa replied:
Is it being set in the environment manually prior to running ecflow workflow after ecflow_client has loaded the definition file to the server.

aerorahul replied:
Ok.  
It would be good to document this in a future PR on how a developer could setup and execute a ecflow workflow.

lgannoaa replied:
Like the rocoto workflow, ecflow is a workflow. There will be a bit learning associated when developer try to use.

lgannoaa committed:
Remove the use of extra include file

lgannoaa committed:
Create new JJob for workflow_monitor

lgannoaa commented:
@aerorahul if you do not have anymore concern on this draft PR, may I convert it to regular PR?

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+##############################################
+# Ensure obsproc input exist
+##############################################
+OBSPROC_GDAS_Trigger=$(compath.py ${envir}/obsproc/${obsproc_ver})/gdas.${PDY}/${cyc}/atmos/gdas.t${cyc}z.updated.status.tm00.bufr_d
+IMSSNOW_GDAS_Trigger=$(compath.py ${envir}/obsproc/${obsproc_ver})/gdas.${PDY}/${cyc}/atmos/gdas.t${cyc}z.imssnow96.grib2
+OBSPROC_GFS_Trigger=$(compath.py ${envir}/obsproc/${obsproc_ver})/gfs.${PDY}/${cyc}/atmos/gfs.t${cyc}z.updated.status.tm00.bufr_d
+IMSSNOW_gfs_Trigger=$(compath.py ${envir}/obsproc/${obsproc_ver})/gfs.${PDY}/${cyc}/atmos/gfs.t${cyc}z.imssnow96.grib2
+COMIN_OBS_gfs=$(compath.py prod/obsproc/${obsproc_ver})/gfs.$PDY/$cyc/atmos/gfs.$cycle.prepbufr
+COMIN_OBS_gdas=$(compath.py prod/obsproc/${obsproc_ver})/gdas.$PDY/$cyc/atmos/gdas.$cycle.prepbufr
```
Can this not be squeezed based on `$CDUMP`?

lgannoaa replied:
This job, JECF_MONITOR control the release of GDAS, GFS, and ENKFGDAS jobs. Maybe using CDUMP will be confuse in here.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+export pid=${pid:-$$}
+export outid=${outid:-"LL$job"}
+
+export DATA=${DATA:-${DATAROOT}/${jobid:?}}
+mkdir -p $DATA
+cd $DATA
```
I don't see any files generated in this script locally into `$DATA`
What is the use of `$DATA` in this job?

lgannoaa replied:
The DATA is a place keeper that in here incase we need. I see there will be needed when we proceed to retro mode. Therefore, I keep it in here.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+export CDATE=${CDATE:-${PDY}${cyc}}
+export CDUMP=${CDUMP:-${RUN:-"gfs"}}
+export COMPONENT=${COMPONENT:-atmos}
+GDATE=$($NDATE -06 $CDATE)
+gPDY=$(echo $GDATE | cut -c1-8)
+gcyc=$(echo $GDATE | cut -c9-10)
```
I don't see the use of these variables anywhere in this job.

lgannoaa replied:
These two are required in verification jobs part of the JECF_MONITOR. However, the requirement changed last Friday (2022/4/8) to separate the verification jobs work into (1) JJobs (2) New design meeting on 2022/04/14. The use of these two parameters will be updated upon the result of the meeting.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+##############################################
+export CDATE=${CDATE:-${PDY}${cyc}}
+export CDUMP=${CDUMP:-${RUN:-"gfs"}}
+export COMPONENT=${COMPONENT:-atmos}
```
Where is `COMPONENT` used in this job?

lgannoaa replied:
It is not use at this time because we are not yet running in retro parallel mode. We may need it in retro mode, therefore I keep it in here as a place holder.

aerorahul replied:
I understand.
However, we are not very good at cleaning up if they are not used later.
I prefer adding when needed and not before unless it is an infrastructure stub.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+    release_gdas="YES"
+    ecflow_client --event release_gdas
+  fi
+  if [ -s "$OBSPROC_GDAS_Trigger" -a -s "$IMSSNOW_gfs_Trigger" -a -s "$COMIN_OBS_gfs" -a $release_gfs = "NO" ]; then
```
Shouldn't this be `OBSPROC_GFS_Trigger`?

lgannoaa replied:
I will modify it.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+release_enkfgdas="NO"
+while [ $ic -le 20 ];
+do
+  if [ -s "$OBSPROC_GDAS_Trigger" -a -s "$IMSSNOW_GDAS_Trigger" -a -s "$COMIN_OBS_gdas" -a $release_gdas = "NO" ]; then
+    release_gdas="YES"
+    ecflow_client --event release_gdas
+  fi
+  if [ -s "$OBSPROC_GDAS_Trigger" -a -s "$IMSSNOW_gfs_Trigger" -a -s "$COMIN_OBS_gfs" -a $release_gfs = "NO" ]; then
+    release_gfs="YES"
+    ecflow_client --event release_gfs
+  fi
+  if [ -s $COMIN_OBS_gdas ]; then
+    release_enkfgdas="YES"
+    ecflow_client --event release_enkfgdas
+  fi
```
I think these can be squeezed based on `CDUMP`

lgannoaa replied:
This job, JECF_MONITOR control the release of GDAS, GFS, and ENKFGDAS jobs. Maybe using CDUMP will be confuse in here.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+# Ensure MetPlus input exist
+##############################################
+release_fg_metpg2o1="NO"
+release_fg_metppcp1="NO"
+COMIN_vppg="/lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/global"
+COMIN_vppg_gdas=${COMIN_vppg}/archive/obs_data/prepbufr/gdas/prepbufr.gdas.${PDYm1}00
+COMIN_vppg_nam_PDY=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDY}        # /nam.t*z.prepbufr.tm*
+COMIN_vppg_nam_PDYm1=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDYm1}    # /nam.t*z.prepbufr.tm*
+COMIN_vppg_nam_PDYm2=${COMIN_vppg}/archive/obs_data/prepbufr/nam/nam.${PDYm2}    # /nam.t*z.prepbufr.tm*
+COMIN_vppg_ccpa=${COMIN_vppg}/archive/obs_data/ccpa_accum24hr/ccpa.${PDYm1}12.24h
+echo "Checking metpg2o1 input for prepbufr file on ${PDYm1}"
+ic=21
+while [ $ic -le 20 ];
+do
+  echo $COMIN_vppg_gdas
+  if [ -s $COMIN_vppg_gdas ]; then
+    cd $COMIN_vppg_nam_PDYm1
+    test_ct_1=`find . -name "nam.t*z.prepbufr.tm*"|wc -l`
+    cd $COMIN_vppg_nam_PDYm2
+    test_ct_2=`find . -name "nam.t*z.prepbufr.tm*"|wc -l`
+    if [ $test_ct_1 -eq 8 ]; then
+      if [ $test_ct_2 -eq 8 ]; then
+        release_fg_metpg2o1="YES"
+      fi
+    fi
+  fi
+  echo "Checking metppcp1 input for prepbufr file on ${PDYm1}"
+  if [ -s $COMIN_vppg_ccpa ]; then
+    release_fg_metppcp1="YES"
+  fi
+  if [ $release_fg_metpg2o1 = "YES" -a $release_fg_metppcp1 = "YES" ]; then
+    ecflow_client --event release_metplus
+    break
+  fi
+  ic=`expr $ic + 1`
+  if [ $ic -gt 20 ]; then
+    exit 99
+  fi
+  sleep 38
+done
```
Can you explain the relevance of this section when there are no `metp` jobs currently running in ecflow?

lgannoaa replied:
These two are required in verification jobs part of the JECF_MONITOR. However, the requirement changed last Friday (2022/4/8) to separate the verification jobs work into (1) JJobs (2) New design meeting on 2022/04/14. 
This area of the code maybe modified after the new design meeting.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,146 @@
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd $DATAROOT
+[[ $KEEPDATA = "NO" ]] && rm -rf $DATA
```
What is in the `DATA` directory?

lgannoaa replied:
The use of a DATA directory is standard in JJobs which is a place holder in case in the this JJob need it. Since we are still not finalize on retro parallel plan, maybe we can keep it here for now.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -15,3 +22,14 @@ fi
+  export PDY=${PDY:-%PDY:%}
+  export CDATE=${PDY}%CYC:%
+fi
+#### Required for archive job to do auto cleanup
+export DATAROOT=/lfs/h2/emc/stmp/%EMC_USER%/RUNDIRS/%PSLOT%/${CDATE}
+mkdir -p $DATAROOT
+if [ -n "%COMPATH:%" ]; then export COMPATH=${COMPATH:-%COMPATH:%}; fi
```
This can be re-introduced/re-evaluated when an archive job is added.
Currently, there is no archive job being run.
Besides, this section seems very specific to the archive job and yet is being executed in `envir-p1.h` which is sourced by all jobs.

lgannoaa replied:
Currently the gfs archive, gdas archive, and enkfgdas earc jobs are all running in my development ecflow workflow.
The archive job need this to ensure cleanup works. It does not impact other jobs.

aerorahul replied:
Since they are not in the repository, I would prefer adding when the jobs are added.
Please reintroduce then.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -15,3 +22,14 @@ fi
+
+export COMROOT=/lfs/h2/emc/ptmp/%EMC_USER%/%PSLOT%/com
+export ROTDIR="$(compath.py -o prod/${NET}/${gfs_ver})"
+if [ -n "%PDY:%" ]; then
+  export PDY=${PDY:-%PDY:%}
+  export CDATE=${PDY}%CYC:%
+fi
```
`PDY` is set in the jjobs with `./setPDY.sh`.
Can you explain why this is being set here?

lgannoaa replied:
This setting sets the PDY on each cycle. That enable the ecflow workflow to run retro parallel.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -2,6 +2,13 @@
+export envir=%ENVIR%
+export MACHINE_SITE=%MACHINE_SITE%
+
+if [ -n "%SENDCANNEDDBN:%" ]; then export SENDCANNEDDBN=${SENDCANNEDDBN:-%SENDCANNEDDBN:%}; fi
```
I don't understand this;
`if [ -n "%SENDCANNEDDBN:%" ]`
The `-n` checks if the string variable length is non-zero.  If it is true, then
`export SENDCANNEDDBN=${SENDCANNEDDBN:-%SENDCANNEDDBN:%};`
If it is already non-zero, should this not already be in the environment space?
What am I missing?

lgannoaa replied:
This is for develop be able to overwrite in case of the need to test. This was used for NCO testing.

aerorahul reviewed (commented):
Please see some comments.

lgannoaa committed:
Add documentation ecflow_dev.md
Remove three metplus and gplot jobs. With new requirement on April 14th, 2022
Modify JECF_MONITOR to remove metplus monitoring.
Ref: 721

lgannoaa committed:
New requirement to modify envir-p1.h to remove archive job related setting
and JECF_MONITOR cleanup
Ref: 721

aerorahul commented on docs/ecflow_dev.md:
```diff
@@ -0,0 +1,22 @@
+Following steps are to help developer install and config ecflow workflow for realtime cycled parallel run
```
Can you please use markdown syntax with itemized steps.
This reads like a text file.

aerorahul commented on jobs/JECF_MONITOR:
```diff
@@ -0,0 +1,89 @@
+  if [ $release_gdas = "YES" -a $release_gfs = "YES" -a $release_enkfgdas = "YES" ]; then
+    break
+  fi
+  ic=`expr $ic + 1`
```
I am pretty sure the jobs/scripts in global-workflow are using `$( )` instead of backticks.  @WalterKolczynski-NOAA am I mistaken.

WalterKolczynski-NOAA replied:
Correct, I removed all backticks from shell scripts. Please use only $( ).

lgannoaa replied:
@aerorahul @WalterKolczynski-NOAA Would you please point me to an example script that meet this new requirement? Looking at the exglobal_forecast.sh in develop branch, I see backticks used.

WalterKolczynski-NOAA replied:
A quick search doesn't show any backticks in that script.

Here is the PR where I removed them all: #526

aerorahul reviewed (commented):
I think most of the changes are good to go in this PR.

lgannoaa committed:
Update document file
Update backticks for new requirement

aerorahul commented on ecf/setup_ecf_links.sh:
```diff
@@ -18,12 +18,31 @@ function link_master_to_fhr(){
 fhrs=($(seq 3 9))
 link_master_to_fhr "jenkfgdas_post" "$fhrs"
 
+# EnKF GDAS earc files
+cd $ECF_DIR/scripts/workflow_manager/cycled/enkfgdas
+echo "Linking enkfgdas/earc ..."
+grps=($(seq 0 8))
+link_master_to_grp "jenkfgdas_emc_earc" "$grps"
```
This would only be executed in `EMC` mode right?

lgannoaa replied:
Yes.

lgannoaa replied:
It was not. I modified this script to exclude NCO mode in linking earc scripts.

lgannoaa committed:
Exclude development mode in linking earc scripts

aerorahul reviewed (approved):
looks good.
Next week, I will try and follow the instructions in `ecflow_dev.md` and see if I can recreate this.

KateFriedman-NOAA commented:
> Next week, I will try and follow the instructions in ecflow_dev.md and see if I can recreate this.

@aerorahul *gentle poke* Have you had a chance to try this? Since I'm assigned to merge this now I'm checking on it's readiness for merging. Let me know, thanks!

WalterKolczynski-NOAA commented:
Where do we stand with this? Is Rahul still supposed to be testing before we merge?

aerorahul commented:
Closing after discussions w/ @lgannoaa
